File: ch1-arraysAndStrings/mohamad_jayyusi/mohamad_jayyusi__1_8.java
/**
 * @author [Mohamad Jayyusi]
 * @email
 * @create date 2021-02-23
 * @modify date 2021-02-23
 * @desc 1.8 - Zero Matrix: Write an algorithm such that if an element in an MxN
 *       matrix is 0, its entire row and column are set to 0.
 */

public class mohamad_jayyusi__1_8 {

    private static final int m = 8;
    private static final int n = 8;
    private static int[][] matrix = new int[m][n];
    // O(n²) a multi dimension array to store the indices where there is a
    // zero in the original matrix
    private static int[][] zeroIndices = new int[m][n];

    public static void main(String[] args) {

        initializeMatrix();
        setZero(0, 0);
        printMatrix();
        // O(n²) additional space
        findZeroIndices();
        zeroMatrix();
        System.out.println();
        printMatrix();
    }

    // O(n²) additional space - O(n²) time complexity
    public static void findZeroIndices() {
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0)
                    zeroIndices[i][j] = 0;
            }
        }
    }

    // O(n²) time complexity
    public static void zeroMatrix() {
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (zeroIndices[i][j] == 0)
                    setRowAndColToZeros(i, j);
            }
        }
    }

    public static void setZero(int m, int n) {
        matrix[m][n] = 0;
    }

    // O(n)
    public static void setRowAndColToZeros(int rowNum, int ColNum) {
        for (int j = 0; j < n; j++) {
            setZero(rowNum, j);
        }
        for (int i = 0; i < m; i++) {
            setZero(i, ColNum);
        }
    }

    public static void initializeMatrix() {
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                matrix[i][j] = 9;
                zeroIndices[i][j] = 1;
            }
        }
    }

    public static void printMatrix() {
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                System.out.print(matrix[i][j]);
            }
            System.out.println();
        }
    }
}

File: ch1-arraysAndStrings/mohamad_jayyusi/mohamad_jayyusi__1_8.py
# /**
#  * @author [Mohamad Jayyusi]
#  * @email 
#  * @create date 2021-02-23
#  * @modify date 2021-02-23
#  * @desc 1.8 - Zero Matrix: Write an algorithm such that if an element in an MxN
#                matrix is 0, its entire row and column are set to 0.
#  */

#set matrix dimension
m = 4
n = 5
#value to fill matrix with
initialValue = 7
matrix = []
zeroIndices = []

#set a cell to 0 - O(1) time complexity
def setZero(i, j):
    matrix[i][j] = 0

#print matrix - O(n) time complexity
def printMatrix():
    for i in range(0,m):
        print(matrix[i])

#create matrix - O(n) time complexity
def createMatrix():
    for i in range(0,m):
        matrix.append([ initialValue ] * n) #O(n) time complexity

#initialize uni dimensional array
#for storing indices of zero cells
#O(n²) time complexity
def zeroIndicesArray():
    for i in range(0,m):
        for j in range(0, n):
          zeroIndices.append(initialValue)

#locate zero cells
#O(n²) time complexity
def findZeroIndices():
    for i in range(0,m):
        for j in range(0, n):
          if matrix[i][j] == 0:
              zeroIndices[i*n+j] = 0

#zero rows and cols
#O(n) time complexity
def setRowAndColToZeros(rowNum, colNum):
    for j in range(0,n):
        setZero(rowNum, j)
    
    for i in range(0,m):
        setZero(i, colNum)

#zero matrix O(n²)
def zeroMatrix():
    for i in range(0,len(zeroIndices)):
        if zeroIndices[i] == 0:
            setRowAndColToZeros(i//n,i%n)

# Solution
createMatrix()
zeroIndicesArray()
printMatrix()
print() # blank line
# setZero(0,0)
setZero(1,1)
# setZero(2,2)
findZeroIndices()
zeroMatrix()
printMatrix()